server/api/utils/lru_cache.py

```diff
@@ -0,0 +1,93 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import collections
+from copy import deepcopy
+
+
+class LRUCache:
+    """LRU cache which solves some of lru_cache from functools deficiencies.
+    Most importantly, the ability to remove/modify a cache value on top
+    of the ability to clear entire cache.
+    We strive to provide similar API to lru_cache.
+    Note this class is used as is instead of using a decorator for performance reasons.
+    If performance is not as important, a decorator can be added.
+    Note this can be used as a decorator, but only with default parameters i.e. maxsize
+    This code is based on https://pastebin.com/LDwMwtp8
+    """
+
+    class CacheInfo:
+        def __init__(self, maxsize: int):
+            self.maxsize = maxsize
+            self.reset()
+
+        def reset(self):
+            self.hits = 0
+            self.misses = 0
+            self.currsize = 0
+
+    def __init__(self, func, maxsize: int = 128):
+        self.cache = collections.OrderedDict()
+        self.func = func
+        self.maxsize = maxsize
+        self._cache_info = self.CacheInfo(maxsize)
+
+    def __call__(self, *args):
+        cache = self.cache
+        if args in cache:
+            self._cache_info.hits += 1
+            cache.move_to_end(args)
+            return cache[args]
+        result = self.func(*args)
+        cache[args] = result
+        self._cache_info.misses += 1
+        if len(cache) > self.maxsize:
+            cache.popitem(last=False)
+        return result
+
+    def cache_info(self) -> CacheInfo:
+        """Get cache statistics"""
+        self._cache_info.currsize = len(self.cache)
+        return deepcopy(self._cache_info)
+
+    def cache_clear(self) -> None:
+        """Remove all values from cache and reset statistics"""
+        self.cache.clear()
+        self._cache_info.reset()
+
+    def cache_remove(self, *args) -> None:
+        """Remove an item from the cache by passing the same args used in the call"""
+        if args in self.cache:
+            self.cache.pop(args)
+
+    def cache_replace(self, value, *args) -> None:
+        """Replace value only if in cache. Use cache_set() if you want to make sure the value is set in the cache"""
+        cache = self.cache
+        if args in cache:
+            cache[args] = value
+            cache.move_to_end(args)
+
+    def cache_set(self, value, *args) -> None:
+        """Set a single value in cache"""
+        cache = self.cache
+        cache[args] = value
+        # TODO: Will this be more efficient if we check if args already in cache first?
```

Comment:
probably, yes. but if it exists don't you want to update the value anyway?
If not, then you can use `cache.setdefault(key, value)`

Reply:
if it exists, I don't need to check if I need to pop a value